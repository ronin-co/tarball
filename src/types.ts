export type Prettify<T> = { [K in keyof T]: T[K] } & {};

export interface TarballInputFile {
  /**
   * The contents of the file.
   */
  contents: Uint8Array;

  /**
   * The name of the file to be included in the tarball.
   */
  name: string;

  /**
   * The date at which the file was last modified.
   *
   * @default new Date(0)
   */
  lastModifiedAt?: Date;
}

export interface CreateTarballOptions {
  /**
   * Whether to compress the tarball using gzip or not
   *
   * @default true
   */
  compress?: boolean;

  /**
   * The timestamp at which the tarball was created.
   *
   * This is required when using gzip compression as without it the integrity
   * hash will change every time.
   *
   * The default value is the package version timestamp.
   */
  timestamp: number;
}

export interface CreateTarballResult<TName extends string> {
  /**
   * The tarball data as a Uint8Array.
   */
  data: Uint8Array<ArrayBuffer>;

  /**
   * The name of the tarball.
   */
  name: TName;
}
