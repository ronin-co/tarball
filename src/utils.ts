import type { TarballInputFile } from '@/src/types';

/**
 * Build the header record for the provided file.
 *
 * @param file - The file to build the header record for.
 *
 * @returns The header record for the provided file.
 */
export const buildHeaderRecord = <T extends TarballInputFile>(file: T): Uint8Array => {
  const lastModifiedAt = file.lastModifiedAt ?? new Date(0);

  const header = new Uint8Array(512);

  const entries = [
    // File name (100 bytes)
    stringToPaddedUint8Array(file.name, 100),
    // File mode (8 bytes)
    stringToZeroFilledUint8Array('644', 8),
    // Owner's numeric user ID (8 bytes)
    stringToZeroFilledUint8Array('0', 8),
    // Group's numeric user ID (8 bytes)
    stringToZeroFilledUint8Array('0', 8),
    // File size in bytes (octal base) (12 bytes)
    stringToZeroFilledUint8Array(file.contents.byteLength.toString(8), 12),
    // Last modification time in numeric Unix time format (octal) (12 bytes)
    stringToZeroFilledUint8Array(
      Math.floor(lastModifiedAt.valueOf() / 1000).toString(8),
      12,
    ),
    // Check sum for header record (8 bytes)
    stringToZeroFilledUint8Array('        ', 8),
    // Link indicator (file type) (1 byte)
    stringToZeroFilledUint8Array('0', 1),
    // Name of linked file (100 bytes)
    stringToPaddedUint8Array('', 100),
    // USTAR indicator "ustar" then NUL (6 bytes)
    stringToPaddedUint8Array('ustar', 6),
    // USTAR version "00" (2 bytes)
    stringToUint8Array('00'),
    // Owner user name (32 bytes)
    stringToPaddedUint8Array('parsable', 32),
    // Owner group name (32 bytes)
    stringToPaddedUint8Array('parsable', 32),
    // Device major number (8 bytes)
    stringToPaddedUint8Array('0', 8),
    // Device minor number (8 bytes)
    stringToPaddedUint8Array('0', 8),
    // Filename prefix (155 bytes)
    stringToPaddedUint8Array('', 155),
  ] satisfies Array<Uint8Array<ArrayBufferLike>>;

  let currentOffset = 0;
  for (const entry of entries) {
    header.set(entry, currentOffset);
    currentOffset += entry.byteLength;
  }

  const checksum = Uint8Array.of(
    ...stringToZeroFilledUint8Array(
      entries
        .reduce((sum, entry) => entry.reduce((subSum, value) => subSum + value, sum), 0)
        .toString(8),
      6,
    ),
    0,
    32,
  );

  header.set(checksum, 148);

  return header;
};

/**
 * Calculate the byte length of the tarball based on the provided files.
 *
 * @param files - The files to include in the tarball.
 *
 * @returns The byte length of the tarball.
 */
export const calculateTarballByteLength = <T extends TarballInputFile>(
  files: Array<T>,
): number => {
  // Data (rounded up to a multiple of 512 bytes) & 512-byte header
  const length = files.reduce(
    (acc, file) => acc + Math.ceil(file.contents.byteLength / 512) * 512 + 512,
    0,
  );

  return (
    length + 1024 // 1024-byte blank records in the tail
  );
};

/**
 * Convert a string to an array of 8-bit unsigned integers.
 *
 * @param value - The string to convert.
 *
 * @returns An array of 8-bit unsigned integers with the provided string.
 */
export const stringToUint8Array = (value: string): Uint8Array => {
  return Uint8Array.from(value.split('').map((x) => x.charCodeAt(0)));
};

/**
 * Convert a string to an array of 8-bit unsigned integers with a fixed length and zero-filled.
 *
 * @param value - The string to convert.
 * @param byteLength - The length of the resulting byte array.
 *
 * @returns An array of 8-bit unsigned integers with the provided string and zero-filled.
 */
export const stringToZeroFilledUint8Array = (
  value: string,
  byteLength: number = value.length,
): Uint8Array => {
  const view = new Uint8Array(byteLength).fill(48, 0, byteLength);
  view.set(stringToUint8Array(value), byteLength - value.length);
  return view;
};

/**
 * Convert a string to an array of 8-bit unsigned integers with a fixed length & padded.
 *
 * @param string - The string to convert.
 * @param byteLength - The length of the resulting byte array.
 *
 * @returns An array of 8-bit unsigned integers with the provided string & padded.
 */
export const stringToPaddedUint8Array = (
  string: string,
  byteLength: number,
): Uint8Array => {
  const view = new Uint8Array(byteLength);
  view.set(stringToUint8Array(string));
  return view;
};
