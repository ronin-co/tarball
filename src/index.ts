import { buildHeaderRecord, calculateTarballByteLength } from '@/src/utils';
import { zip as gzip } from 'gzip-js';

import type {
  CreateTarballOptions,
  CreateTarballResult,
  Prettify,
  TarballInputFile,
} from '@/src/types';

/**
 * Create a tarball from a list of files
 *
 * @param name - The name of the tarball
 * @param files - The list of files to include in the tarball
 * @param [options] - The options for creating the tarball
 *
 * @returns An object containing the tarball data and file name.
 */
export const createTarball = <
  TName extends string,
  TFiles extends Array<TarballInputFile>,
>(
  name: TName,
  files: TFiles,
  options?: CreateTarballOptions,
): Prettify<CreateTarballResult<TName>> => {
  const { compress = true, timestamp } = options ?? {};

  const tarballByteLength = calculateTarballByteLength(files);
  const tarballByteArr = new Uint8Array(tarballByteLength);

  let offset = 0;
  for (const file of files) {
    tarballByteArr.set(buildHeaderRecord(file), offset);
    tarballByteArr.set(new Uint8Array(file.contents), offset + 512);

    offset += Math.ceil(file.contents.byteLength / 512) * 512 + 512;
  }

  if (compress) {
    const gzipTarball = gzip(tarballByteArr, {
      name,
      timestamp,
    });

    return {
      name,
      data: new Uint8Array(gzipTarball),
    };
  }

  return {
    name,
    data: tarballByteArr,
  };
};
