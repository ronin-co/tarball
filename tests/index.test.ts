import { describe, expect, test } from 'bun:test';

import { createTarball } from '@/src/index';

/**
 * Get the integrity hash of a given data
 *
 * @param data - A Uint8Array containing the data to hash
 * @param algorithm - The algorithm to use for hashing (Default: `'SHA-512'`)
 *
 * @returns A promise that resolves to the integrity hash string of the provided data
 */
const getIntegrityHash = async (
  data: Uint8Array,
  algorithm: 'SHA-1' | 'SHA-256' | 'SHA-384' | 'SHA-512' = 'SHA-512',
): Promise<string> => {
  const digest = await crypto.subtle.digest(algorithm, data);
  const digestByteArr = new Uint8Array(digest);
  const hash = btoa(String.fromCharCode(...digestByteArr));
  return `${algorithm.toLowerCase().replace('-', '')}-${hash}`;
};

// Note: All tarballs created for these tests require a fixed timestamp to ensure
// the integrity hash remains consistent across test runs.
const CREATED_AT_TIMESTAMP = new Date('2025-01-01T00:00:00.000Z').getTime();

describe('create', () => {
  test('a basic tarball', async () => {
    const tarball = createTarball(
      'basic.tar.gz',
      [
        {
          name: 'hello-world.txt',
          contents: new Uint8Array(Buffer.from('Hello, world!', 'utf-8')),
        },
      ],
      {
        timestamp: CREATED_AT_TIMESTAMP,
      },
    );
    const tarballHash = await getIntegrityHash(tarball);

    expect(tarballHash).toMatchSnapshot();
  });

  test('an empty tarball', async () => {
    const tarball = createTarball('empty.tar.gz', [], {
      timestamp: CREATED_AT_TIMESTAMP,
    });
    const tarballHash = await getIntegrityHash(tarball);

    expect(tarballHash).toMatchSnapshot();
  });

  test('an uncompressed tarball', async () => {
    const tarball = createTarball('basic.tar.gz', [], {
      compress: false,
      timestamp: CREATED_AT_TIMESTAMP,
    });
    const tarballHash = await getIntegrityHash(tarball);

    expect(tarballHash).toMatchSnapshot();
  });
});
