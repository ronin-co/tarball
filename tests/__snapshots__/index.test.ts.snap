// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`create a basic tarball 1`] = `"sha512-ZgtIFb/6PHovvz1IGB/8IUHmvQhnJXZcyxwpq365SzJbXGqq5pH1YBzVtMMGlse4ZgVQLRLoHp4NQZNEoblnKQ=="`;

exports[`create an empty tarball 1`] = `"sha512-TbZfkhWIV5aDktJFZIchaMbYctV8kqCHMy2LgGKO0K0tfM0yXRNRvYitFlqim2LXb9MVoZuQSJOc7lcWeL/D1w=="`;

exports[`create an uncompressed tarball 1`] = `"sha512-jvtPc8VlU1HEROsQkjDFVtOeLHYk6cEavJ4/tLm5JUIYzFCFtFSpaY0IXPqSGYSR8HpyO+RXStxwYXtz6wtkYQ=="`;
